### DISCUSSION TITLE
Can I hide object parameters?

### DISCUSSION THREAD

**Discussion by schwidom** [LogtalkDotOrg/logtalk3#204] [Q&A] [answered]
I tried to hide an object parameter by using a kindof stub for that:

```
:- object( o001(_T_)).

:- private( punifytest/2).

punifytest( A, A).

:- public( pu1/1).
:- public( pu2/1).

pu1(A) :- punifytest(A,B), (A,B) = _T_.
pu2(B) :- punifytest(A,B), (A,B) = _T_.


:- end_object.

:- object( o001, extends( o001(_))).

:- end_object.
```

When I use the parameterized version I get the expected answer:

```
?- o001(_)::(pu1(A), pu2(B)).
A = B.
```

But when I use the stub the binding does not exist:

```
?- o001::(pu1(A), pu2(B)).
true.
```

Is this a bug?

**Comment by pmoura** [comment]
There's no bug here. Your definition of the `o001` object doesn't bind the parent object parameter. The `o001::(pu1(A), pu2(B))` goal is translated to `o001(_)::pu1(A), o001(_)::pu2(B)`:

```text
?- o001(_)::pu1(A), o001(_)::pu2(B).
true.
```

The two anonymous variables are different variables and thus there's no variable sharing between the two goals. Note that the `o001(_)::(pu1(A), pu2(B))` goal is translated into the `o001(T)::pu1(A), o001(T)::pu2(B)` conjunction and thus parameter sharing occurs in this case.

**Reply by schwidom** [reply]
Ok, and is there a possibility to create the object parameter variable without being in the parameter list? I am searching for a possibility to hide this variable.

**Reply by pmoura** [reply]
Parameter variables are logic variables. But only parameter variables are scoped to the whole parametric object/category. There's no mechanism that would allow creating a parameter variable without it being referenced in the object/category identifier. Curious about your use case...

**Comment by schwidom** [comment]
> Curious about your use case...

In OO languages you have the builder pattern where you can initialize an object step by step by calling methods which in turn return the same object of the same type so you are able to continue to build. What I want is to do  the same with logical variables. This object parameter could be an assoc which I create with o::new. New creates the keys in the assoc. And afterwards I call other predicates which unify the values o::(new, p1(1), p2(3)). I want to make sure that the object parameter is a free variable in the beginning.
The concrete usecase is that I want to assemble SAT formulae. In the assoc then I will have variables which are bound to specific SAT formulae and several predicates provide the variables in the assoc and return the formulae itself so that I can later on create a final SAT formula like: o::(new, p1(SAT1), p2(SAT2), p3(SAT3),vars(VARS)), TERM= SAT1+SAT2*SAT3, sat(TERM), labeling(VARS) ... .

All these variables which I have in the assoc could indeed also be single object variables. But I don't need them outside. And the amount of variables is growing over time. The only thing I can do is to put the object creation in a separate predicate.

**Reply by pmoura** [reply]
Maybe define a wrapper object that hides the parametric object?

**Reply by schwidom** [reply]
I tried it (on top of the discussion: :- object( o001, extends( o001(_))). ) but this does not work. Is there another possiblilty to wrap this?

**Reply by pmoura** [reply, marked as answer]
Not by extending but by _using_ the parametric object from within the wrapper object.

**Reply by schwidom** [reply]
Ok that will work, thanks. But it seems to me it is only restricted to the inner part of objects. Is there a possibility to use this alias mechanism outside of objects? Maybe also for normal prolog terms?

**Reply by pmoura** [reply]
Not currently. I do have draft support for using [`uses/1`](https://logtalk.org/manuals/refman/directives/uses_1.html) directives in _user_. But it would only make a difference for top-level queries.
